internal/wasm/sdk/internal/wasm/bindings.go

```diff
@@ -4,93 +4,196 @@
 
 package wasm
 
-// #include <stdlib.h>
-//
-// extern void opa_println(void *context, int32_t addr);
-// extern void opa_abort(void *context, int32_t addr);
-// extern int32_t opa_builtin0(void *context, int32_t builtin_id, int32_t ctx);
-// extern int32_t opa_builtin1(void *context, int32_t builtin_id, int32_t ctx, int32_t arg0);
-// extern int32_t opa_builtin2(void *context, int32_t builtin_id, int32_t ctx, int32_t arg0, int32_t arg1);
-// extern int32_t opa_builtin3(void *context, int32_t builtin_id, int32_t ctx, int32_t arg0, int32_t arg1, int32_t arg2);
-// extern int32_t opa_builtin4(void *context, int32_t builtin_id, int32_t ctx, int32_t arg0, int32_t arg1, int32_t arg2, int32_t arg3);
-import "C"
-
 import (
-	"unsafe"
-
-	wasm "github.com/wasmerio/go-ext-wasm/wasmer"
+	"bytes"
+	"context"
+	"encoding/json"
+	"strconv"
+	"time"
+
+	"github.com/bytecodealliance/wasmtime-go"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/metrics"
+	"github.com/open-policy-agent/opa/topdown"
+	"github.com/open-policy-agent/opa/topdown/builtins"
 )
 
-func opaFunctions(imports *wasm.Imports) (*wasm.Imports, error) {
-	imports, err := imports.AppendFunction("opa_println", opa_println, C.opa_println)
-	if err != nil {
-		return nil, err
+func opaFunctions(dispatcher *builtinDispatcher, store *wasmtime.Store) []*wasmtime.Extern {
+
+	i32 := wasmtime.NewValType(wasmtime.KindI32)
+
+	externs := []*wasmtime.Extern{
+		wasmtime.NewFunc(store, wasmtime.NewFuncType([]*wasmtime.ValType{i32}, nil), opaAbort).AsExtern(),
+		wasmtime.NewFunc(store, wasmtime.NewFuncType([]*wasmtime.ValType{i32, i32}, []*wasmtime.ValType{i32}), dispatcher.Call).AsExtern(),
+		wasmtime.NewFunc(store, wasmtime.NewFuncType([]*wasmtime.ValType{i32, i32, i32}, []*wasmtime.ValType{i32}), dispatcher.Call).AsExtern(),
+		wasmtime.NewFunc(store, wasmtime.NewFuncType([]*wasmtime.ValType{i32, i32, i32, i32}, []*wasmtime.ValType{i32}), dispatcher.Call).AsExtern(),
+		wasmtime.NewFunc(store, wasmtime.NewFuncType([]*wasmtime.ValType{i32, i32, i32, i32, i32}, []*wasmtime.ValType{i32}), dispatcher.Call).AsExtern(),
+		wasmtime.NewFunc(store, wasmtime.NewFuncType([]*wasmtime.ValType{i32, i32, i32, i32, i32, i32}, []*wasmtime.ValType{i32}), dispatcher.Call).AsExtern(),
 	}
 
-	imports, err = imports.AppendFunction("opa_abort", opa_abort, C.opa_abort)
-	if err != nil {
-		return nil, err
+	return externs
+}
+
+func opaAbort(caller *wasmtime.Caller, args []wasmtime.Val) ([]wasmtime.Val, *wasmtime.Trap) {
+
+	data := caller.GetExport("memory").Memory().UnsafeData()[args[0].I32():]
```

Comment:
https://pkg.go.dev/github.com/bytecodealliance/wasmtime-go#Memory.UnsafeData

> This is not a safe method to call, hence the "unsafe" in the name. The byte slice returned from this function is not managed by the Go garbage collector. You need to ensure that `m`, the original `Memory`, lives longer than the `[]byte` returned.

I suppose that's something to keep in mind if weird things start to happen 💭

Reply:
We _should_ be fine since the memory is referenced by the VM so as long as the VM struct is around alive, the memory should be alive. Definitely keeping in mind if we start running into unexpected panics though.